File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
          int m = mat.length, n = mat[0].length;
        int[][] res = new int[m][n];
         //bfs
        Queue<int[]> q = new LinkedList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    q.offer(new int[]{i, j});
                } else {
                    res[i][j] = Integer.MAX_VALUE;
                }
            }
        }
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            for (int[] dir : dirs) {
                int x = cur[0] + dir[0], y = cur[1] + dir[1]; 
                if (x < 0 || x >= m || y < 0 || y >= n)continue;  
                if(res[x][y] <= res[cur[0]][cur[1]] + 1) continue;
                res[x][y] = res[cur[0]][cur[1]] + 1;
                q.offer(new int[]{x, y});
            }
        }
        
        
        return res;
    }
}

File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        q = deque((i, j) for j in range(n) for i in range(m) if mat[i][j] == 0)
        mp = [[0 if mat[i][j] == 0 else float("-inf") for j in range(n)] for i in range(m)]
        while q:
            i, j = q.popleft()
            for ci, cj in [(i + 1, j), (i - 1, j), (i, j + 1), (i, j - 1)]:
                if 0 <= ci < m and 0 <= cj < n and mp[ci][cj] == float("-inf"):
                    mp[ci][cj] = mp[i][j] + 1
                    q.append((ci, cj))
        return mp